armi/reactor/converters/axialExpansionChanger.py

```diff
@@ -506,22 +506,31 @@ def _determineLinked(componentA, componentB):
         and isinstance(componentA, type(componentB))
         and (componentA.getDimension("mult") == componentB.getDimension("mult"))
     ):
-        idA, odA = (
-            componentA.getCircleInnerDiameter(cold=True),
-            componentA.getBoundingCircleOuterDiameter(cold=True),
-        )
-        idB, odB = (
-            componentB.getCircleInnerDiameter(cold=True),
-            componentB.getBoundingCircleOuterDiameter(cold=True),
-        )
+        try:
```

Comment:
I'd rather that we just explicitly check the instances / types of the components are `Unshaped` and then handle this directly. I don't mind the try / except but that should result in a failure because we should enforce that for other components that the developers do implement `getCircleInnerDiameter` and `getBoundingCircleOuterDiameter`.

Just my thought here so we don't hide future problems leading to some complex / time consuming debugging sessions. Also, does this work if we specify the target components from in #777. Just trying to make sure I understand when this method is called.

Reply:
@jakehader I tend to agree with your sentiment - see the "Note:" in my PR description. 

The blanket try/except was the result of a conversation between @keckler and I. Chris - what are your thoughts on changing the try/except to just handle the unshaped component explicitly? Off the top of my head we could do something like this:

```python
if (
    (componentA.containsSolidMaterial() and componentB.containsSolidMaterial())
    and isinstance(componentA, type(componentB))
    and (componentA.getDimension("mult") == componentB.getDimension("mult"))
):
    if isinstance(componentA, type(UnshapedComponent):
        runLog.warning(
            f"Components {componentA} and {componentB} are UnshapedComponents "
            "and do not have 'getCircleInnerDiameter' or getBoundingCircleOuterDiameter methods; "
            "nor is it physical to do so. Instead of crashing and raising an error, they are going to be assumed to not be linked.",
            single=True,
        )
        linked = False
    else:
        idA, odA = (
            componentA.getCircleInnerDiameter(cold=True),
            componentA.getBoundingCircleOuterDiameter(cold=True),
        )
       ...
```

Reply:
I personally prefer it the way that it is currently implemented. My reasoning for this is that it will not always make sense for every component type to have the `getCircleInnerDiameter` and/or `getBoundingCircleOuterDiameter` methods. Right now we are thinking only about `UnshapedComponent`s, but I think there are other components that also wouldn't have those methods, like any 3D shapes (which we probably wouldn't want to be linked ever anyways).

So, in my opinion, I think the current `try` / `except` makes more sense because it is more general. We give a warning, so what more can we really do without being overly restrictive?

Reply:
Is it too bulky/unmaintainable to do something like this for future component types?
```python 
if (
      isinstance(componentA, type(UnshapedComponent)
      or isinstance(componentA, type(NewFutureComponent1)
      or isinstance(componentA, type(NewFutureComponent2)
    ):
        runLog.warning( ...
``` 
This could definitely be cleaner, but is a potential jumping off point for discussions. 

But I see your point @keckler . I guess it comes down to two options:
1. we leave the try/except with a runLog.warning and rely on users to recognize it/check the warnings at the end of an ARMI stdout.
2. we do the above and have ARMI throw a Runtime Error if the new component isn't supported. Then either they can modify the source as needed and create a PR or create an issue for one of us to fix it. 

@john-science do you have a preference? Or maybe a secret third option?

Reply:
I would like the framework to fail fast and fail early, so I am afraid of having generic exception handling. We have done this in the past and in many ways results in a lot of time to debug issues or it masks bugs that may already exist today! In the spirit of keeping this PR to one solution for one issue I am still recommending this. I am interested to hear your take on this though @keckler and of course @john-science.

Reply:
I definitely understand what you're saying. 

The main reasoning behind my opinion is that I feel the current implementation is more in line with ARMI being a "framework" that anybody could use for any use-case, as opposed to a tool that is tailored specifically to our needs at TP. In this context, I think adding a clause to an if-statement for the specific case of an `UnshapedComponent` is just very overly-specific and not in the spirit of a "framework", because it `UnshapedComponent`s are just one use-case out of many potential use-cases.

That is my opinion, but I do not need to die on this sword. So I won't argue if we go with @jakehader 's suggestion, which it sounds like we will be doing. It still fixes my current problem, so I can't complain.

Reply:
I understand. I think we have been bitten in the past with over-generalizing too soon and that is my main concern here. I am not sure which issues that will pop-up and masking them with unbounded `try/excepts` has caused issues in the past. @albeanth, if you can implement my suggestion then I will approve.

Reply:
Addressed https://github.com/terrapower/armi/pull/778/commits/73469189bb1feec05fd54a73caad769abc55de0a